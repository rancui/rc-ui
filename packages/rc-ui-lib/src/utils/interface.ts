import type { CSSProperties } from 'react';

export interface BaseTypeProps {
  style?: CSSProperties;
  className?: string;
  children?: React.ReactNode;
}

/** 指定挂载的节点 */
export type TeleportType = HTMLElement | (() => HTMLElement) | null;
/** 带有displayName的组件 */
export type WithDisplayNameReactElement = React.ReactElement & { type: { displayName: string } };
