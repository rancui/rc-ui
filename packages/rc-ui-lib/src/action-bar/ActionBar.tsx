import React, { useContext, useMemo, useRef } from 'react';
import classnames from 'classnames';
import { ActionBarProps } from './PropsType';
import ActionBarContext from './ActionBarContext';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { usePlaceholder } from '../hooks/use-placeholder';

const ActionBar: React.FC<ActionBarProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('action-bar', prefixCls);
  const root = useRef<HTMLDivElement>(null);

  const children = useMemo(() => React.Children.toArray(props.children), [props.children]);

  const renderPlaceholder = usePlaceholder(root, bem);

  const renderActionBar = () => (
    <ActionBarContext.Provider value={{ parent: { children } }}>
      <div
        ref={root}
        className={classnames(props.className, bem(), {
          'rc-safe-area-bottom': props.safeAreaInsetBottom,
        })}
        style={props.style}
      >
        {React.Children.toArray(props.children)
          .filter(Boolean)
          .map((child: React.ReactElement, index: number) => {
            return React.cloneElement(child, {
              index,
            });
          })}
      </div>
    </ActionBarContext.Provider>
  );

  return props.placeholder ? renderPlaceholder(renderActionBar) : renderActionBar();
};

ActionBar.defaultProps = {
  safeAreaInsetBottom: true,
  placeholder: false,
};

export default ActionBar;
