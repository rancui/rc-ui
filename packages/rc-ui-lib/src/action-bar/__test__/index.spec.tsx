import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import Icon from '../../icon';
import ActionBar from '..';
import { sleep } from '../../../tests/utils';
import { mockGetBoundingClientRect } from '../../utils/dom/mock';

describe('ActionBar', () => {
  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  it('should allow to disable safe-area-inset-bottom prop', () => {
    const { container } = render(
      <ActionBar safeAreaInsetBottom={false}>
        <ActionBar.Icon icon="chat-o" text="客服" />
        <ActionBar.Icon icon="cart-o" text="购物车" />
        <ActionBar.Icon icon="shop-o" text="店铺" />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render placeholder element when using placeholder prop', async () => {
    const restore = mockGetBoundingClientRect({ height: 50 });
    const { container } = render(
      <ActionBar safeAreaInsetBottom={false} placeholder>
        <ActionBar.Icon icon="chat-o" text="客服" />
        <ActionBar.Icon icon="cart-o" text="购物车" />
        <ActionBar.Icon icon="shop-o" text="店铺" />
      </ActionBar>,
    );

    await sleep(0);
    expect(container).toMatchSnapshot();
    restore();
  });

  it('basic usage with ActionBar.Icon', async () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon="chat-o" text="客服" />
        <ActionBar.Icon icon="cart-o" text="购物车" />
        <ActionBar.Icon icon="shop-o" text="店铺" />
        <ActionBar.Button type="danger" text="立即购买" />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when using badge prop in ActionBarIcon', async () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon="chat-o" text="客服" badge={{ dot: true }} />
        <ActionBar.Icon icon="cart-o" text="购物车" badge={{ content: 5 }} />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when color prop in ActionBarIcon', async () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon="chat-o" color="#ee0a24" text="客服" />
        <ActionBar.Icon icon="cart-o" text="购物车" />
        <ActionBar.Icon icon="star" color="#ff5000" text="店铺" />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when icon prop in ActionBarIcon', async () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon="chat-o" color="#ee0a24" text="客服" />
        <ActionBar.Icon icon={<Icon name="shop-o" />} text="购物车" />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('when icon prop in ActionBarIcon is not isValidElement', async () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon={123} color="#ee0a24" text="客服" />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when iconClass prop in ActionBarIcon', async () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon="chat-o" iconClass="icon-class" text="客服" />
        <ActionBar.Icon icon="cart-o" text="购物车" />
      </ActionBar>,
    );
    expect(container.querySelector('.icon-class')).toBeTruthy();
  });

  it('should render correctly when iconPrefix prop in ActionBar.Icon', async () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon="chat-o" iconPrefix="my-icon" text="客服" />
        <ActionBar.Icon icon="cart-o" text="购物车" />
      </ActionBar>,
    );
    expect(container.querySelector('.my-icon-chat-o')).toBeTruthy();
  });

  it('should emit click event in ActionBar.Icon', async () => {
    const onClick = jest.fn();
    const { container } = render(
      <ActionBar>
        <ActionBar.Icon icon="chat-o" text="客服" onClick={onClick} />
        <ActionBar.Icon icon="cart-o" text="购物车" />
      </ActionBar>,
    );
    const icon = container.querySelector('.van-icon-chat-o');
    fireEvent.click(icon);
    expect(onClick).toHaveBeenCalled();
  });

  it('should render correctly when using type prop and text prop in ActionBarButton', () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Button type="warning" text="加入购物车" />
        <ActionBar.Button type="danger" text="立即购买" />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when using icon prop in ActionBarButton', () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Button type="warning" text="加入购物车" icon="shop-o" />
        <ActionBar.Button type="danger" text="立即购买" icon={<Icon name="chat-o" />} />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when using color prop in ActionBarButton', () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Button type="warning" text="加入购物车" color="red" />
        <ActionBar.Button
          type="danger"
          text="立即购买"
          color="linear-gradient(to right, #ffd01e, #ff8917)"
        />
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('when ActionBarButton does not wrapperd by ActionBar', () => {
    const { container } = render(
      <div>
        <ActionBar.Button type="warning" text="加入购物车" />
        <ActionBar.Button type="danger" text="立即购买" />
      </div>,
    );
    expect(container).toMatchSnapshot();
  });

  it('when ActionBarButton has children', () => {
    const { container } = render(
      <ActionBar>
        <ActionBar.Button type="warning">加入购物车</ActionBar.Button>
        <ActionBar.Button type="danger">立即购买</ActionBar.Button>
      </ActionBar>,
    );
    expect(container).toMatchSnapshot();
  });
});
