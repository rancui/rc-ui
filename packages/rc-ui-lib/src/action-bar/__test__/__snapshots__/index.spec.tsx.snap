// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionBar basic usage with ActionBar.Icon 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-chat-o"
      />
      客服
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-cart-o"
      />
      购物车
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-shop-o"
      />
      店铺
    </div>
    <button
      class="rc-action-bar-button rc-action-bar-button--danger rc-action-bar-button--last rc-action-bar-button--first rc-button rc-button--danger rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          立即购买
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`ActionBar should allow to disable safe-area-inset-bottom prop 1`] = `
<div>
  <div
    class="rc-action-bar"
  >
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-chat-o"
      />
      客服
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-cart-o"
      />
      购物车
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-shop-o"
      />
      店铺
    </div>
  </div>
</div>
`;

exports[`ActionBar should render correctly when color prop in ActionBarIcon 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-chat-o"
        style="color: rgb(238, 10, 36);"
      />
      客服
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-cart-o"
      />
      购物车
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-star"
        style="color: rgb(255, 80, 0);"
      />
      店铺
    </div>
  </div>
</div>
`;

exports[`ActionBar should render correctly when icon prop in ActionBarIcon 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-chat-o"
        style="color: rgb(238, 10, 36);"
      />
      客服
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon"
      >
        <i
          class="rc-badge__wrapper van-icon van-icon-shop-o"
        />
      </div>
      购物车
    </div>
  </div>
</div>
`;

exports[`ActionBar should render correctly when using badge prop in ActionBarIcon 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-chat-o"
      >
        <div
          class="rc-badge rc-badge--top-right rc-badge--dot rc-badge--fixed"
        />
      </div>
      客服
    </div>
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-cart-o"
      >
        <div
          class="rc-badge rc-badge--top-right rc-badge--fixed"
        >
          5
        </div>
      </div>
      购物车
    </div>
  </div>
</div>
`;

exports[`ActionBar should render correctly when using color prop in ActionBarButton 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <button
      class="rc-action-bar-button rc-action-bar-button--warning rc-action-bar-button--first rc-button rc-button--warning rc-button--large"
      style="color: rgb(255, 255, 255); background: red; border-color: red;"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          加入购物车
        </span>
      </div>
    </button>
    <button
      class="rc-action-bar-button rc-action-bar-button--danger rc-action-bar-button--last rc-button rc-button--danger rc-button--large"
      style="color: rgb(255, 255, 255); border: 0px;"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          立即购买
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`ActionBar should render correctly when using icon prop in ActionBarButton 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <button
      class="rc-action-bar-button rc-action-bar-button--warning rc-action-bar-button--first rc-button rc-button--warning rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <i
          class="rc-badge__wrapper rc-button__icon van-icon van-icon-shop-o"
        />
        <span
          class="rc-button__text"
        >
          加入购物车
        </span>
      </div>
    </button>
    <button
      class="rc-action-bar-button rc-action-bar-button--danger rc-action-bar-button--last rc-button rc-button--danger rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <i
          class="rc-badge__wrapper van-icon van-icon-chat-o"
        />
        <span
          class="rc-button__text"
        >
          立即购买
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`ActionBar should render correctly when using type prop and text prop in ActionBarButton 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <button
      class="rc-action-bar-button rc-action-bar-button--warning rc-action-bar-button--first rc-button rc-button--warning rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          加入购物车
        </span>
      </div>
    </button>
    <button
      class="rc-action-bar-button rc-action-bar-button--danger rc-action-bar-button--last rc-button rc-button--danger rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          立即购买
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`ActionBar should render placeholder element when using placeholder prop 1`] = `
<div>
  <div
    class="rc-action-bar__placeholder"
    style="height: 50px;"
  >
    <div
      class="rc-action-bar"
    >
      <div
        class="rc-action-bar-icon"
        role="button"
        tabindex="0"
      >
        <div
          class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-chat-o"
        />
        客服
      </div>
      <div
        class="rc-action-bar-icon"
        role="button"
        tabindex="0"
      >
        <div
          class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-cart-o"
        />
        购物车
      </div>
      <div
        class="rc-action-bar-icon"
        role="button"
        tabindex="0"
      >
        <div
          class="rc-badge__wrapper rc-action-bar-icon__icon van-icon van-icon-shop-o"
        />
        店铺
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionBar when ActionBarButton does not wrapperd by ActionBar 1`] = `
<div>
  <div>
    <button
      class="rc-action-bar-button rc-action-bar-button--warning rc-button rc-button--warning rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          加入购物车
        </span>
      </div>
    </button>
    <button
      class="rc-action-bar-button rc-action-bar-button--danger rc-button rc-button--danger rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          立即购买
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`ActionBar when ActionBarButton has children 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <button
      class="rc-action-bar-button rc-action-bar-button--warning rc-action-bar-button--first rc-button rc-button--warning rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          加入购物车
        </span>
      </div>
    </button>
    <button
      class="rc-action-bar-button rc-action-bar-button--danger rc-action-bar-button--last rc-button rc-button--danger rc-button--large"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          立即购买
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`ActionBar when icon prop in ActionBarIcon is not isValidElement 1`] = `
<div>
  <div
    class="rc-action-bar rc-safe-area-bottom"
  >
    <div
      class="rc-action-bar-icon"
      role="button"
      tabindex="0"
    >
      客服
    </div>
  </div>
</div>
`;
